api/workload/v2/workload.go

```diff
@@ -0,0 +1,219 @@
+func WithAddr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		addr:     DefaultAgentAddress,
+		watcher:  watcher,
+		ctx:      ctx,
+		cancelFn: cancel,
+		backoff:  newBackoff(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
+	c.wg.Add(1)
+	go c.run()
+	return nil
+}
+
+func (c *X509SVIDClient) run() {
+	defer c.wg.Done()
+
+	conn := c.newConn()
+	if conn == nil {
+		return
+	}
+	c.watcher.OnConnection(true)
+	defer c.watcher.OnConnection(false)
+	defer conn.Close()
+
+	c.watch(conn)
+}
+
+// establishes a new persistent connection, returns nil if a client can't be created within the context deadline
+func (c *X509SVIDClient) newConn() *grpc.ClientConn {
+	for {
+		conn, err := grpc.DialContext(c.ctx, c.addr, grpc.WithInsecure())
+		if err != nil {
+			if done := c.handleError(err); done {
+				return nil
+			}
+			continue
+		}
+		c.backoff.Reset()
+		return conn
+	}
+}
+
+// handles an error, applies backoff, and returns true if the context has been canceled
+func (c *X509SVIDClient) handleError(err error) (done bool) {
+	if status.Code(err) == codes.Canceled {
+		return true
+	}
+	c.watcher.OnError(err)
+	select {
+	case <-time.After(c.backoff.Duration()):
+		return false
+	case <-c.ctx.Done():
+		return true
+	}
+}
+
+// creates single watch loop and returns whether we should continue watching
+func (c *X509SVIDClient) watch(conn *grpc.ClientConn) {
+	for {
+		stream, err := c.newX509SVIDStream(conn)
+		if err != nil {
+			if done := c.handleError(err); done {
+				return
+			}
+			continue
+		}
+		if err := c.handleX509SVIDStream(stream); err != nil {
+			if done := c.handleError(err); done {
+				return
+			}
+		}
+	}
+}
+
+func (c *X509SVIDClient) newX509SVIDStream(conn *grpc.ClientConn) (workload.SpiffeWorkloadAPI_FetchX509SVIDClient, error) {
+	workloadClient := workload.NewSpiffeWorkloadAPIClient(conn)
+	header := metadata.Pairs("workload.spiffe.io", "true")
+	grpcCtx := metadata.NewOutgoingContext(c.ctx, header)
+	return workloadClient.FetchX509SVID(grpcCtx, &workload.X509SVIDRequest{})
+}
+
+func (c *X509SVIDClient) handleX509SVIDStream(stream workload.SpiffeWorkloadAPI_FetchX509SVIDClient) error {
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		svids, err := protoToX509SVIDs(resp)
+		if err != nil {
+			return err
```

Comment:
returning an error here will cause backoff on an otherwise healthy stream. should this just call OnError() and `continue`?

More importantly, if the stream does not Recv() everything, it will stay active and we'll essentially leak the stream until the context passed into `FetchX509SVID` is canceled. I'd recommend `watch` create a cancellable context from c.ctx that it passes into newX509SvidStream that it can cancel when `handleX509SVIDStream` returns just to make sure we don't leak streams.

Reply:
ah, good calll

Reply:
i need to add some tests for this